src/runtime/HalideBuffer.h

```diff
@@ -553,7 +553,8 @@ class Buffer {
 
     /** A pointer to the element with the lowest address. If all
      * strides are positive, equal to the host pointer. */
-    T *begin() const {
+    // @{
+    T *begin() {
```

Comment:
I worry that removing the const qualifier from the is going to break something. This class is a bit schizophrenic about what `const Buffer<T>` means. The metadata is certainly immutable, but what about the data? What's the difference between `const Buffer<T>` and `const Buffer<const T>`? I had thought that the former meant immutable metadata and mutable data, but operator() follows the stl style of making the data immutable when the container is.

I guess what I'm saying is that we could also just mark for_each_value as const. It doesn't change the metadata, and it will already either pass T& or const T& to the lambda depending on whether T has a const qualifier.

Reply:
Yeah, you have a good point. Let me try making that simplifying assumption and see if it addresses what I need addressed. (If so, then `begin`/`end` would go back as they were, with additional comments added about why const methods returning apparently-but-not-necessarily-nonconst-pointers is intended and safe.)

(The impetus for this change is various inside-google image classes that don't have this dual const distinction, and so lots of code typically wraps a `const ImageType>` with a  `const Buffer<T>` on the assumption the latter is a fully read-only variant; most of those really should be `const Buffer<const T>`, which I'm currently pondering how to gradually fix without having to boil the ocean.)